README.md
# bubble-project


ChangeLog.md
# Changelog for bubble-project

## Unreleased changes


test/Spec.hs
import qualified Test.Lib as Lib
import qualified Test.Gui as Gui

main :: IO ()
main = Lib.test >> Gui.test

--main = putStrLn "Test suite not yet implemented"


test/Test/Gui.hs
module Test.Gui where

import Test.QuickCheck
import Gui

prop_cutString :: String -> Bool
prop_cutString str = if length str > 1 then 1 + length (cutString (length str - 1) 1 str) == length str else null (cutString (length str - 1) 1 str)

test :: IO ()
test = putStrLn "Test cutString function" >> quickCheck prop_cutString 

test/Test/Lib.hs
module Test.Lib where

import Test.QuickCheck
--import Test.QuickCheck.All
--import Test.QuickCheck.Modifiers
import Lib
import System.Random


-- test distance
prop_distance :: ((Float, Float),Int) -> ((Float, Float),Int) -> Bool
prop_distance ((x1,y1),c1) ((x2,y2),c2) = 
    distance ((x1,y1),c1) ((x2,y2),c2) == distance ((y1,x1),c1) ((y2,x2),c2) && distance ((x1,y1),c1) ((x2,y2),c2) == distance ((x2,y2),c2) ((x1,y1),c1)
    && distance ((x1,y1),c1) ((x2,y2),c2) >=0

prop_distanceT :: ((Float, Float),Int) -> (Float, Float) -> Bool
prop_distanceT ((x1,y1),c) (x2,y2) = 
    distanceT ((x1,y1),c) (x2,y2) == distanceT ((y1,x1),c) (y2,x2) && distanceT ((x1,y1),c) (x2,y2) == distanceT ((x2,y2),c) (x1,y1)
    && distanceT ((x1,y1),c) (x2,y2) >= 0

prop_emptyBalls :: [((Float, Float), Int)] -> Int -> Bool
prop_emptyBalls balls rand = not (null (emptyBalls balls (mkStdGen rand)))

prop_colorBall :: Int -> Bool
prop_colorBall rand = fst (colorBall (mkStdGen rand)) <= 7 && fst (colorBall (mkStdGen rand)) >= 1

angleTable1 :: [Float]
angleTable1 = [-70..70]
angleTable2 :: [Float]
angleTable2 = map (+0.5) [-70..69]
angleTable :: [Float]
angleTable = angleTable1 ++ angleTable2

tablePomoc :: [Float]
tablePomoc = [0..]
tabc :: [Int]
tabc = [1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,71,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,
    1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7,1,2,3,4,5,6,7]

instance Arbitrary Action where
    arbitrary = elements [NoA, RightA, LeftA]


instance Show Game where
    --show gameBalls = " "
    --show gameBall = " "
    --show gameOver = " "
    --show gameScore = " "
    --show gameAction = " "
    --show gameAngle = " "
    --show gameShoot  = " "
    --show pomoc = " "
    --show gameTable = " "
    --show gameRandom = " "
    --show defaultBalls = " "

instance Arbitrary Game where
    arbitrary = do
        ballx <- arbitrary
        bally <- arbitrary
        ballc <- chooseInt (1,7)
        ballspos <-sublistOf table1 
        ballsc <- sublistOf tabc
        ameOver <- elements [True,False]
        ameScore <- arbitrary
        ameAction <- elements [NoA, RightA, LeftA]
        ameAngle <- arbitrary
        ameShoot <- elements [True,False]
        omoc <- elements tablePomoc
        ameLicznik <- arbitrary
        ameTable <- sublistOf table1
        rand <- choose (1,500)
        ameRecord <- arbitrary
        return $ Game ((ballx,bally),ballc) (ameBalls ballspos ballsc) ameOver ameScore ameAction ameAngle ameShoot omoc ameLicznik ameTable (mkStdGen rand) ameRecord efaultBalls
        where
            ameBalls ballspos ballsc = zip ballspos ballsc
            efaultBalls = [((1,1),1),((2,1),1),((3,1),2),((4,1),2),((5,1),3),((6,1),1),((7,1),4),((8,1),4),((9,1),5),((10,1),5),((11,1),5),((12,1),6)]
            


prop_angleMove :: Action -> Float -> Int -> Bool
prop_angleMove act ang rand 
    | act == LeftA = if ang <= 70 then gameAngle (angleMove (Game ((6.5,24),1) [((1,1),1)] False 0 act ang False 0 0 table1 (mkStdGen rand) 10 [((1,1),1)])) == ang+0.5
        else gameAngle (angleMove (Game ((6.5,24),1) [((1,1),1)] False 0 act ang False 0 0 table1 (mkStdGen rand) 10 [((1,1),1)])) == ang
    | act == RightA = if ang >= -70 then gameAngle (angleMove (Game ((6.5,24),1) [((1,1),1)] False 0 act ang False 0 0 table1 (mkStdGen rand) 10 [((1,1),1)])) == ang-0.5
        else gameAngle (angleMove (Game ((6.5,24),1) [((1,1),1)] False 0 act ang False 0 0 table1 (mkStdGen rand) 10 [((1,1),1)])) == ang
    | otherwise = gameAngle (angleMove (Game ((6.5,24),1) [((1,1),1)] False 0 act ang False 0 0 table1 (mkStdGen rand) 10 [((1,1),1)])) == ang

prop_angleMoveAddRow :: [((Float,Float),Int)] -> Int -> Bool
prop_angleMoveAddRow balls licz
    | rem licz 6 == 5 = length (gameBalls (Game ((6.5,24),1) balls False 0 NoA 0 False 0 licz table1 (mkStdGen 200) 10 [((1,1),1)])) < length (gameBalls (angleMove (Game ((6.5,24),1) balls False 0 NoA 0 False 0 licz table1 (mkStdGen 200) 10 [((1,1),1)])))
    | otherwise = length (gameBalls (Game ((6.5,24),1) balls False 0 NoA 0 False 0 licz table1 (mkStdGen 200) 10 [((1,1),1)])) == length (gameBalls (angleMove (Game ((6.5,24),1) balls False 0 NoA 0 False 0 licz table1 (mkStdGen 200) 10 [((1,1),1)])))

gameHelp :: Game -> Game
gameHelp (Game ((x,y),c) balls over score act ang shoot pom licz _ rand re def) = Game ((x,y),c) balls over score act ang shoot pom licz acT rand re def 
    where 
        acT = if length (activeT balls table1) < length (activeT def table2) then activeT def table1 else activeT balls table2

prop_changeBalls :: Game -> Bool 
prop_changeBalls game = 
    length (changeBalls (gameBall (gameHelp game)) (gameBalls (gameHelp game))) == 1 + length (gameBalls (gameHelp game)) 
        || length (changeBalls (gameBall (gameHelp game)) (gameBalls (gameHelp game))) <= (-2) + length (gameBalls (gameHelp game))




test :: IO ()
test = do
    putStrLn "Test distance function" >> quickCheck prop_distance
    putStrLn "Test distanceT function" >> quickCheck prop_distanceT
    putStrLn "Test emptyBalls function" >> quickCheck prop_emptyBalls
    putStrLn "Test colorBall function" >> quickCheck prop_colorBall
    putStrLn "Test angleMove function" >> quickCheck (withMaxSuccess 1000 prop_angleMove) 
    putStrLn "Test angleMove with add row" >> quickCheck prop_angleMoveAddRow
    putStrLn "Test changeBalls function" >> quickCheck (withMaxSuccess 6000 prop_changeBalls)


src/Gui.hs
module Gui where

import Graphics.Gloss
import Graphics.Gloss.Interface.IO.Game as G
import Lib

gameBackground :: Color
gameBackground = white
gameWindow :: Display      --gameWindow = InWindow "Bubble" (260,500) (100, 100)
gameWindow = InWindow "Bubble" (325,620) (100, 100)

keys :: Event -> Game -> IO Game
keys (EventKey (SpecialKey KeyLeft) G.Down _ _) game = if not $ gameShoot game then return (changeAction LeftA game) else return game
keys (EventKey (SpecialKey KeyRight) G.Down _ _) game = if not $ gameShoot game then return (changeAction RightA game) else return game
keys (EventKey (SpecialKey KeyDown) G.Down _ _) game = if not $ gameShoot game then return (changeAction NoA game) else return game
keys (EventKey (SpecialKey KeyUp) G.Down _ _) game = if not $ gameShoot game && not (gameOver game) 
                        then return (changeShoot True game) else return game
keys (EventKey (SpecialKey KeySpace) G.Down _ _) game = 
    if gameOver game && (notElem 22 (map (snd . fst) (gameBalls game)) && not (null (gameBalls game))) then return (startingGame True game) else return game
keys (EventKey (SpecialKey KeyEnter) G.Down _ _) game = 
    if gameOver game then return (startingGameDef True game) else return game
keys (EventKey (SpecialKey KeyEsc) G.Down _ _) game = if gameOver game then return game else return (endGame True game)
keys _ game = return game

graphicsGame :: Game -> IO Picture
graphicsGame game = return ( pictures $ scoreGraphics
                            ++ recordGraphics
                            ++ fmap bubbleGraphics balls
                            ++ [bubbleGraphics (gameBall game)]
                            ++ scoreTextGraphics
                            ++ recordTextGraphics
                            ++ [gameLine (gameAngle game)]
                            ++ [gameOverLine]
                            ++ overGraphics
                            )
    where
        gameLine alpha = color black $ scale 1 (-1) $ line [(6.5*20-130,24*20-250),((6.5-2*sin x)*20-130,(24-2*cos x)*20-250)]
            where
                x = alpha * pi / 180
        gameOverLine = color red $ scale 1 (-1) $ line [(0.5*20-130,21.5*20-250),(12.5*20-130,21.5*20-250)]
        fillRectangle color' (tx, ty) (w, h) =  color color' $ 
                                                    scale 1 (-1) $ 
                                                    translate (tx * 20 - 130) (ty * 20 - 250) $ 
                                                    rectangleSolid w h
        balls = gameBalls game
        bubbleGraphics ::((Float,Float),Int) -> Picture
        bubbleGraphics ((x,y),col)
            | col == 1 = color red $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | col == 2 = color blue $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | col == 3 = color azure $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | col == 4 = color green $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | col == 5 = color rose $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | col == 6 = color orange $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
            | otherwise = color cyan  $ scale 1 (-1) $ translate (x*20 - 130) (y*20 - 250) $ circleSolid 9
        scoreGraphics = [color red $ translate 100 (-290) $ scale 0.25 0.25 $ text (show (gameScore game))]
        recordGraphics = [color red $ translate 100 275 $ scale 0.25 0.25 $ text (show (gameRecord game))]
        scoreTextGraphics = [color blue $ translate 40 (-285) $ scale 0.1 0.1 $ text "Score: "]
        recordTextGraphics = [color blue $ translate 35 280 $ scale 0.1 0.1 $ text "Record: "]
        overGraphics = if gameOver game then [color red $ translate (-120) 20 $ scale 0.25 0.25 $ text "Koniec gry :("
                                     ,  color black $ translate (-110) (-25) $ scale 0.1 0.1 $ text "Press Enter to start default game" 
                                     ,  color black $ translate (-130) (-50) $ scale 0.1 0.1 $ text "Press Space to start (only if correct file)"] 
                                    else [ fillRectangle black (6.5, 0) (260, 20)
                                        , fillRectangle black (6.5, 25) (260, 20)
                                        , fillRectangle black (0, 12.5) (20, 500)
                                        , fillRectangle black (13, 12.5) (20, 500) ]

convIntToString :: Int -> String
convIntToString = show
convFloatToString :: Float -> String
convFloatToString = show

convLIntToLString :: [Int] -> [String]
convLIntToLString = map convIntToString
convLFloatToLString :: [Float] -> [String]
convLFloatToLString = map convFloatToString

cutString :: Int -> Int -> String -> String
cutString _ _ "" = ""
cutString n m l = take n l ++ cutString n m (drop (n+m) l)


step :: Float -> Game -> IO Game
step _ game = do
    writeFile "record.txt" (show $ gameRecord game)
    writeFile "bx.txt" $ cutString (lx-1) 1 (unlines $ convLFloatToLString (map (fst . fst) (gameBalls game))) 
    writeFile "by.txt" $ cutString (ly-1) 1 (unlines $ convLFloatToLString (map (snd . fst) (gameBalls game)))
    writeFile "bc.txt" $ cutString (lc-1) 1 (unlines $ convLIntToLString (map snd (gameBalls game))) 
    if gameOver game then return game else
        return (Game ((x,y),c) balls over score action angle shoot pom licz activ rand re def)
    where
        lx = length (unlines $ convLFloatToLString (map (fst . fst) (gameBalls game)))
        ly = length (unlines $ convLFloatToLString (map (snd . fst) (gameBalls game)))
        lc = length (unlines $ convLIntToLString (map snd (gameBalls game)))
        over = ourGameOver balls
        action = gameAction game
        Game ((x,y),c) balls _ score _ angle shoot pom licz activ rand re _ = move game
        def = defaultBalls game
        


f1 :: String -> Int
f1 = read
f2 :: String -> Float
f2 = read

outFunc :: IO ()
outFunc = do
    x <- readFile "bubblex.txt"
    y <- readFile "bubbley.txt"
    c <- readFile "bubblec.txt"
    bx <- readFile "bx.txt"
    by <- readFile "by.txt"
    bc <- readFile "bc.txt"
    record <- fmap f1 (readFile "record.txt")
    let ab = zip (zip (map f2 . words $ bx) (map f2 . words $ by)) (map f1 . words $ bc)
    let a = zip (zip (map f2 . words $ x) (map f2 . words $ y)) (map f1 . words $ c)
    playIO gameWindow gameBackground 35 (startGame False ab record a) graphicsGame keys step


src/Lib.hs
module Lib where

import Data.List
import Data.Ord
import System.Random

---------------------------------------------------------------
columns :: Float
rows :: Float
columns = 13
rows = 25

data Action = NoA | RightA | LeftA deriving (Eq, Show)
data Game = Game {    gameBall :: ((Float,Float),Int)
                    , gameBalls :: [((Float,Float),Int)]
                    , gameOver :: Bool
                    , gameScore :: Int
                    , gameAction :: Action
                    , gameAngle :: Float
                    , gameShoot :: Bool
                    , pomoc :: Float
                    , gameLicznik :: Int
                    , gameTable :: [(Float,Float)]
                    , gameRandom :: StdGen
                    , gameRecord :: Int
                    , defaultBalls :: [((Float,Float),Int)]
                    }

startGame :: Bool -> [((Float,Float),Int)] -> Int -> [((Float,Float),Int)] -> Game
startGame start balls re def = Game {   gameBall = ((6.5,24),2)
                    , gameBalls = balls
                    , gameOver = not start
                    , gameScore = 0
                    , gameAction = NoA
                    , gameAngle = 0
                    , gameShoot = False
                    , pomoc = 0
                    , gameLicznik = 0
                    , gameTable = if length (activeT balls table1) < length (activeT balls table2) then activeT balls table1
                                else activeT balls table2
                    , gameRandom = mkStdGen 100
                    , gameRecord = re
                    , defaultBalls = def
                    }

startingGame :: Bool -> Game -> Game
startingGame over (Game _ balls _ _ _ _ _ _ _ acT rand re def) = Game ((6.5,24),col) balls (not over) 0 NoA 0 False 0 0 acT nextRand re def
    where 
        (col,nextRand) = colorBall rand

startingGameDef :: Bool -> Game -> Game
startingGameDef over (Game _ _ _ _ _ _ _ _ _ _ rand re def) = Game ((6.5,24),col) def (not over) 0 NoA 0 False 0 0 acT nextRand re def
    where 
        acT = if length (activeT def table1) < length (activeT def table2) then activeT def table1 else activeT def table2
        (col,nextRand) = colorBall rand

endGame :: Bool -> Game -> Game
endGame end (Game ball balls _ score act ang shoot _ _ acT rand re def) = Game ball balls end score act ang shoot 0 0 acT rand re def

table1 :: [(Float,Float)]
table1 = [(x,y) | x <- [1,2,3,4,5,6,7,8,9,10,11,12], y <- [1,3,5,7,9,11,13,15,17,19,21]] 
    ++ [(x,y) | x <- [1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5,9.5,10.5,11.5], y <- [2,4,6,8,10,12,14,16,18,20,22]]
table2 :: [(Float,Float)]
table2 = [(x,y) | x <- [1,2,3,4,5,6,7,8,9,10,11,12], y <- [2,4,6,8,10,12,14,16,18,20,22]] 
    ++ [(x,y) | x <- [1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5,9.5,10.5,11.5], y <- [1,3,5,7,9,11,13,15,17,19,21]]

nonActiveT :: [((Float,Float),Int)] -> [(Float,Float)]
nonActiveT = map fst

activeT :: [((Float,Float),Int)] -> [(Float,Float)] -> [(Float,Float)]
activeT balls = filter ( `notElem` nonActiveT balls)

-------------------------------- keyboard -----------------------------
changeAction :: Action -> Game -> Game
changeAction LeftA (Game pos balls over score _ ang shoot pom licz at rand re def)
    | ang >= 70 = Game pos balls over score NoA ang shoot pom licz at rand re def
    | otherwise = Game pos balls over score LeftA ang shoot pom licz at rand re def
changeAction RightA (Game pos balls over score _ ang shoot pom licz at rand re def)
    | ang <= -70 = Game pos balls over score NoA ang shoot pom licz at rand re def
    | otherwise = Game pos balls over score RightA ang shoot pom licz at rand re def
changeAction NoA (Game pos balls over score _ ang shoot pom licz at rand re def) = Game pos balls over score NoA ang shoot pom licz at rand re def

changeShoot :: Bool -> Game -> Game
changeShoot shoot (Game ball balls over score act ang _ pom licz at rand re def)
    | shoot = Game ball balls over score NoA ang shoot pom licz at rand re def
    | otherwise = Game ball balls over score act ang shoot pom licz at rand re def
-------------------------------------------------------------------------

moveA :: Float -> Float -> Float
moveA change angle = angle + change 

angleMove :: Game -> Game
angleMove (Game ball balls over score act ang shoot pom licz at rand re def) = Game ball nextBalls over score act nextAng shoot pom nextLicz nextAT rand re def
    where
        nextLicz = if rem licz 6 == 5 then licz+1 else licz
        nextBalls = if rem licz 6 == 5 then addBallsRow balls rand else balls
        nextAT
            | rem licz 6 /= 5 = at
            | length (activeT nextBalls table1) < length (activeT nextBalls table2) = activeT nextBalls table1
            | otherwise = activeT nextBalls table2
        nextAng
            | act == LeftA && ang < 70 = moveA 0.5 ang
            | act == RightA && ang > (-70) = moveA (-0.5) ang
            | otherwise = ang

distance :: ((Float, Float),Int) -> ((Float, Float),Int) -> Float
distance ((x0,y0),_) ((x,y),_) = sqrt $ (x0-x)**2+(y0-y)**2

distanceT :: ((Float, Float),Int) -> (Float, Float) -> Float
distanceT ((x0,y0),_) (x,y) = sqrt $ (x0-x)**2+(y0-y)**2

distanceElemList :: ((Float,Float),Int) -> [((Float,Float),Int)] -> [Float]
distanceElemList ((x0,y0),c) = map (distance ((x0,y0),c))

distanceElemListT :: ((Float,Float),Int) -> [(Float,Float)] -> [Float]
distanceElemListT ((x0,y0),c) = map (distanceT ((x0,y0),c))

distanceList :: ((Float, Float), Int) -> [((Float, Float), Int)] -> Float
distanceList point list = minimum $ distanceElemList point list

minimalElemIndex :: [Float] -> Int
minimalElemIndex tab = fst . minimumBy (comparing snd) $ zip [0..] tab

shootMove :: Game -> Game
shootMove (Game ((x0,y0),c) balls over score act angle shoot pom licz at rand re def)
    | distanceList ((x0,y0),c) balls < 0.9 = Game ((6.5,24),col) nextBalls over (score + nextScore) act angle False 0 nextLicz nextAT nextRand nextRe def
    | y<=1 = Game ((6.5,24),col) nextBalls over (score + nextScore) act angle False 0 nextLicz nextAT nextRand nextRe def
    | x0<1 = Game ((a1,b1),c) balls over score act angle shoot (pom+1) licz at rand re def
    | x0>12 = Game ((a2,b2),c) balls over score act angle shoot (pom+1) licz at rand re def
    | otherwise = Game ((x,y),c) balls over score act angle shoot pom licz at rand re def
    where
        nextLicz
            | null (changeBalls (ballpos,c) balls) = 0
            | otherwise = licz + 1
            where
                ballpos = at !! minimalElemIndex (distanceElemListT ((x0,y0),c) at)
        nextRe = if score + nextScore > re then score + nextScore else re
        --nextOver = ourGameOver nextBalls
        (col,nextRand) = colorBall rand
        nextBalls = emptyBalls (changeBalls (ballpos,c) balls) rand
            where ballpos = at !! minimalElemIndex (distanceElemListT ((x0,y0),c) at)
        nextAT
            | null (changeBalls (ballpos,c) balls) = if length (activeT nextBalls table1) < length (activeT nextBalls table2)
                                 then activeT nextBalls table1 else activeT nextBalls table2
            | changeBalls (ballpos, c) balls == (ballpos, c) : balls = delete (at !! minimalElemIndex (distanceElemListT ((x0, y0), c) at)) at
            | length (activeT tableBalls table1) < length (activeT tableBalls table2) = activeT tableBalls table1
            | otherwise = activeT tableBalls table2
            where 
                ballpos = at !! minimalElemIndex (distanceElemListT ((x0,y0),c) at)
                tableBalls = filterBalls (checkDelateBalls (ballpos,c) (((ballpos,c),True): zip balls (repeat False)))
        nextScore
            | l <= 2 = 0
            | otherwise = (-1) + l + length (filterToScore (delateFreeBalls b (noColorBalls b)))
            where
                ballpos = at !! minimalElemIndex (distanceElemListT ((x0,y0),c) at)
                l = length (filterToScore (checkDelateBalls (ballpos,c) (((ballpos,c),True): zip balls (repeat False))))
                b = filterBalls (checkDelateBalls (ballpos,c) (((ballpos,c),True) : zip balls (repeat False)))
        (a1,b1) = (1,y0-0.2*cos alpha)
        (a2,b2) = (12,y0-0.2*cos alpha)
        (x,y) = (x0-0.2*((-1)**pom)*sin alpha,y0-0.2*cos alpha)
        alpha = angle * pi / 180

move :: Game -> Game
move (Game ((x0,y0),c) balls over score act angle shoot pom licz at rand re def)
    | shoot = shootMove (Game ((x0,y0),c) balls over score act angle shoot pom licz at rand re def)
    | otherwise = angleMove (Game ((x0,y0),c) balls over score act angle shoot pom licz at rand re def)

neighbourTrue :: Int -> [(((Float,Float),Int),Bool)] -> ((Float,Float),Int) -> (((Float,Float),Int),Bool)
neighbourTrue c balls ((x,y),col)
    | c /= col = (((x,y),col),False)
    | (((x,y),col),True) `elem` balls = (((x,y),col),True)
    | elem (((x-1,y),c),True) balls || elem (((x+1,y),c),True) balls || elem (((x-0.5,y+1),c),True) balls || elem (((x+0.5,y+1),c),True) balls || elem (((x-0.5,y-1),c),True) balls || elem (((x+0.5,y-1),c),True) balls = (((x,y),c),True)
    | otherwise = (((x,y),c),False)

checkDelateBalls :: ((Float,Float),Int) -> [(((Float,Float),Int),Bool)] -> [(((Float,Float),Int),Bool)]
checkDelateBalls ((x,y),c) ballsTrue
    | ballsTrue == map (neighbourTrue c ballsTrue) (map fst ballsTrue) = ballsTrue
    | otherwise = checkDelateBalls ((x,y),c) (map (neighbourTrue c ballsTrue) (map fst ballsTrue))

filterBalls :: [(((Float,Float),Int),Bool)] -> [((Float,Float),Int)]
filterBalls balls = map fst (filter (not . snd) balls)

filterToScore :: [(((Float,Float),Int),Bool)] -> [((Float,Float),Int)]
filterToScore balls = map fst (filter snd balls)

changeBalls :: ((Float,Float),Int) -> [((Float,Float),Int)] -> [((Float,Float),Int)]
changeBalls ((x,y),c) balls = if l > 2 then filterBalls (delateFreeBalls b (noColorBalls b)) else ((x,y),c) : balls
    where
        b = filterBalls (checkDelateBalls ((x,y),c) ((((x,y),c),True) : zip balls (repeat False)))
        l = length (filterToScore (checkDelateBalls ((x,y),c) ((((x,y),c),True) : zip balls (repeat False))))

noColorBall :: ((Float,Float),Int) -> ((Float,Float),Bool)
noColorBall ((x,y),_)
    | y == 1 = ((x,y),False)
    | otherwise = ((x,y),True)

noColorBalls :: [((Float,Float),Int)] -> [((Float,Float),Bool)]
noColorBalls = map noColorBall

neighbourFalse :: [((Float,Float),Bool)] -> ((Float,Float),Int) -> ((Float,Float),Bool)
neighbourFalse balls ((x,y),_)
    | ((x,y),False) `elem` balls = ((x,y),False)
    | elem ((x-1,y),False) balls || elem ((x+1,y),False) balls || elem ((x-0.5,y+1),False) balls || elem ((x+0.5,y+1),False) balls || elem ((x-0.5,y-1),False) balls || elem ((x+0.5,y-1),False) balls = ((x,y),False)
    | otherwise = ((x,y),True)

etykateFreeBalls :: [((Float,Float),Int)] -> [((Float,Float),Bool)] -> [((Float,Float),Bool)]
etykateFreeBalls balls ballsTrue
    | map (neighbourFalse ballsTrue) balls == ballsTrue = ballsTrue
    | otherwise = etykateFreeBalls balls (map (neighbourFalse ballsTrue) balls)

changeForm :: ((Float,Float),Int) -> ((Float,Float),Bool) -> (((Float,Float),Int),Bool)
changeForm ball (_,b) = (ball,b)

changeForms :: [((Float,Float),Int)] -> [((Float,Float),Bool)] -> [(((Float,Float),Int),Bool)]
changeForms = zipWith changeForm

delateFreeBalls :: [((Float,Float),Int)] -> [((Float,Float),Bool)] -> [(((Float,Float),Int),Bool)]
delateFreeBalls balls ballsFalse = changeForms balls (etykateFreeBalls balls ballsFalse)

colorBall :: StdGen -> (Int,StdGen)
colorBall = randomR (1,7)

ourGameOver :: [((Float,Float),Int)] -> Bool
ourGameOver balls
    | 22 `elem` map (snd . fst) balls = True
    | otherwise = False

colorBalls :: StdGen -> [Int]
colorBalls = take 12 . unfoldr (Just . randomR (1, 7))

translateBall :: ((Float,Float),Int) -> ((Float,Float),Int)
translateBall ((x,y),c) = ((x,y+1),c)

translateBalls :: [((Float,Float),Int)] -> [((Float,Float),Int)]
translateBalls = map translateBall

addBallsRow :: [((Float,Float),Int)] -> StdGen -> [((Float,Float),Int)]
addBallsRow balls rand
    | length (activeT (translateBalls balls) table1) < length (activeT (translateBalls balls) table2) = zip (filter (\x -> snd x == 1) table1) (colorBalls rand) ++ translateBalls balls
    | otherwise = zip (filter (\x -> snd x == 1) table2) (colorBalls rand) ++ translateBalls balls

emptyBalls :: [((Float, Float), Int)] -> StdGen -> [((Float, Float), Int)]
emptyBalls balls rand
    | null balls = addBallsRow balls rand
    | otherwise = balls
